User Overview Analysis

In [3]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np

In [80]:
import sys
import os
sys.path.append(os.path.abspath(os.path.join('../scripts')))
from data_selector import *
from data_cleaner import *
from data_visualizer import *
from data_outlier_handler import OutlierHandler

Load Data

In [8]:
# reading the csv file
missing_values = ["n/a", "na", "undefined"]
tellco_data = pd.read_csv("C:/Users/USER/Desktop/KAIM/TellCo-Data-Analysis/Data/TellCo_dataset.csv", na_values=missing_values)
tellco_data

Bearer Id            Start  Start ms              End  End ms  \
0       1.311448e+19   4/4/2019 12:01     770.0  4/25/2019 14:35   662.0   
1       1.311448e+19   4/9/2019 13:04     235.0   4/25/2019 8:15   606.0   
2       1.311448e+19   4/9/2019 17:42       1.0  4/25/2019 11:58   652.0   
3       1.311448e+19   4/10/2019 0:31     486.0   4/25/2019 7:36   171.0   
4       1.311448e+19  4/12/2019 20:10     565.0  4/25/2019 10:40   954.0   
...              ...              ...       ...              ...     ...   
149996  7.277826e+18   4/29/2019 7:28     451.0   4/30/2019 6:02   214.0   
149997  7.349883e+18   4/29/2019 7:28     483.0  4/30/2019 10:41   187.0   
149998  1.311448e+19   4/29/2019 7:28     283.0  4/30/2019 10:46   810.0   
149999  1.311448e+19   4/29/2019 7:28     696.0  4/30/2019 10:40   327.0   
150000           NaN              NaN       NaN              NaN     NaN   

        Dur. (ms)          IMSI  MSISDN/Number          IMEI  \
0       1823652.0  2.082014e+14   3.366496e+10  3.552121e+13   
1       1365104.0  2.082019e+14   3.368185e+10  3.579401e+13   
2       1361762.0  2.082003e+14   3.376063e+10  3.528151e+13   
3       1321509.0  2.082014e+14   3.375034e+10  3.535661e+13   
4       1089009.0  2.082014e+14   3.369980e+10  3.540701e+13   
...           ...           ...            ...           ...   
149996    81230.0  2.082022e+14   3.365069e+10  3.548311e+13   
149997    97970.0  2.082019e+14   3.366345e+10  3.566051e+13   
149998    98249.0  2.082017e+14   3.362189e+10  3.572121e+13   
149999    97910.0  2.082021e+14   3.361962e+10  8.618620e+13   
150000        NaN           NaN            NaN           NaN   

           Last Location Name  ...  Youtube DL (Bytes)  Youtube UL (Bytes)  \
0       9.16456699548519E+015  ...        1.585461e+07        2.501332e+06   
1                     L77566A  ...        2.024740e+07        1.911173e+07   
2                     D42335A  ...        1.972566e+07        1.469958e+07   
3                     T21824A  ...        2.138812e+07        1.514664e+07   
4                     D88865A  ...        1.525938e+07        1.896287e+07   
...                       ...  ...                 ...                 ...   
149996                D20434A  ...        1.619167e+07        1.176343e+07   
149997                D10223C  ...        1.387723e+07        8.288284e+06   
149998                T51102A  ...        2.266051e+07        1.855903e+06   
149999                L88342B  ...        8.817106e+06        8.305402e+06   
150000                    NaN  ...        1.163407e+07        1.100941e+07   

        Netflix DL (Bytes)  Netflix UL (Bytes)  Gaming DL (Bytes)  \
0             8.198936e+06        9.656251e+06       2.780823e+08   
1             1.833841e+07        1.722713e+07       6.087501e+08   
2             1.758779e+07        6.163408e+06       2.295846e+08   
3             1.399465e+07        1.097942e+06       7.995382e+08   
4             1.712458e+07        4.152180e+05       5.277072e+08   
...                    ...                 ...                ...   
149996        1.788370e+07        1.967816e+07       5.266097e+08   
149997        1.935015e+07        2.129315e+07       6.268931e+08   
149998        9.963942e+06        5.065760e+06       5.535395e+08   
149999        3.322253e+06        1.317259e+07       3.525370e+08   
150000        1.162685e+07        1.100175e+07       4.220447e+08   

        Gaming UL (Bytes)  Other DL (Bytes)  Other UL (Bytes)  \
0            1.434415e+07      1.717444e+08      8.814393e+06   
1            1.170709e+06      5.269042e+08      1.505514e+07   
2            3.956300e+05      4.106926e+08      4.215763e+06   
3            1.084972e+07      7.490399e+08      1.279728e+07   
4            3.529801e+06      5.507095e+08      1.391032e+07   
...                   ...               ...               ...   
149996       9.197207e+06      3.264510e+06      1.348742e+07   
149997       4.735033e+06      7.121804e+08

In [5]:
tellco_data.head(5)

Bearer Id            Start  Start ms              End  End ms  \
0  1.311448e+19   4/4/2019 12:01     770.0  4/25/2019 14:35   662.0   
1  1.311448e+19   4/9/2019 13:04     235.0   4/25/2019 8:15   606.0   
2  1.311448e+19   4/9/2019 17:42       1.0  4/25/2019 11:58   652.0   
3  1.311448e+19   4/10/2019 0:31     486.0   4/25/2019 7:36   171.0   
4  1.311448e+19  4/12/2019 20:10     565.0  4/25/2019 10:40   954.0   

   Dur. (ms)          IMSI  MSISDN/Number          IMEI  \
0  1823652.0  2.082014e+14   3.366496e+10  3.552121e+13   
1  1365104.0  2.082019e+14   3.368185e+10  3.579401e+13   
2  1361762.0  2.082003e+14   3.376063e+10  3.528151e+13   
3  1321509.0  2.082014e+14   3.375034e+10  3.535661e+13   
4  1089009.0  2.082014e+14   3.369980e+10  3.540701e+13   

      Last Location Name  ...  Youtube DL (Bytes)  Youtube UL (Bytes)  \
0  9.16456699548519E+015  ...          15854611.0           2501332.0   
1                L77566A  ...          20247395.0          19111729.0   
2                D42335A  ...          19725661.0          14699576.0   
3                T21824A  ...          21388122.0          15146643.0   
4                D88865A  ...          15259380.0          18962873.0   

   Netflix DL (Bytes)  Netflix UL (Bytes)  Gaming DL (Bytes)  \
0           8198936.0           9656251.0        278082303.0   
1          18338413.0          17227132.0        608750074.0   
2          17587794.0           6163408.0        229584621.0   
3          13994646.0           1097942.0        799538153.0   
4          17124581.0            415218.0        527707248.0   

   Gaming UL (Bytes)  Other DL (Bytes)  Other UL (Bytes)  Total UL (Bytes)  \
0         14344150.0       171744450.0         8814393.0        36749741.0   
1          1170709.0       526904238.0        15055145.0        53800391.0   
2           395630.0       410692588.0         4215763.0        27883638.0   
3         10849722.0       749039933.0        12797283.0        43324218.0   
4          3529801.0       550709500.0        13910322.0        38542814.0   

   Total DL (Bytes)  
0       308879636.0  
1       653384965.0  
2       279807335.0  
3       846028530.0  
4       569138589.0  

[5 rows x 55 columns]

In [7]:
# reading the field descriptions file
tellco_field_desc_data = pd.read_excel("C:/Users/USER/Desktop/KAIM/TellCo-Data-Analysis/Data/Field Descriptions.xlsx")
tellco_field_desc_data

Fields  \
0                                  bearer id   
1                                  Dur. (ms)   
2                                      Start   
3                                   Start ms   
4                                        End   
5                                     End ms   
6                                   Dur. (s)   
7                                       IMSI   
8                              MSISDN/Number   
9                                       IMEI   
10                        Last Location Name   
11                           Avg RTT DL (ms)   
12                           Avg RTT UL (ms)   
13                   Avg Bearer TP DL (kbps)   
14                   Avg Bearer TP UL (kbps)   
15               TCP DL Retrans. Vol (Bytes)   
16               TCP UL Retrans. Vol (Bytes)   
17                       DL TP < 50 Kbps (%)   
18            50 Kbps < DL TP < 250 Kbps (%)   
19             250 Kbps < DL TP < 1 Mbps (%)   
20                        DL TP > 1 Mbps (%)   
21                       UL TP < 10 Kbps (%)   
22             10 Kbps < UL TP < 50 Kbps (%)   
23            50 Kbps < UL TP < 300 Kbps (%)   
24                      UL TP > 300 Kbps (%)   
25                           HTTP DL (Bytes)   
26                           HTTP UL (Bytes)   
27                 Activity Duration DL (ms)   
28                 Activity Duration UL (ms)   
29                               Dur. (ms).1   
30                      Handset Manufacturer   
31                              Handset Type   
32           Nb of sec with 125000B < Vol DL   
33     Nb of sec with 1250B < Vol UL < 6250B   
34  Nb of sec with 31250B < Vol DL < 125000B   
35            Nb of sec with 37500B < Vol UL   
36    Nb of sec with 6250B < Vol DL < 31250B   
37    Nb of sec with 6250B < Vol UL < 37500B   
38             Nb of sec with Vol DL < 6250B   
39             Nb of sec with Vol UL < 1250B   
40                   Social Media DL (Bytes)   
41                   Social Media UL (Bytes)   
42                        YouTube DL (Bytes)   
43                        YouTube UL (Bytes)   
44                        Netflix DL (Bytes)   
45                        Netflix UL (Bytes)   
46                         Google DL (Bytes)   
47                         Google UL (Bytes)   
48                          Email DL (Bytes)   
49                          Email UL (Bytes)   
50                         Gaming DL (Bytes)   
51                         Gaming UL (Bytes)   
52                                  Other DL   
53                                  Other UL   
54                          Total DL (Bytes)   
55                          Total UL (Bytes)   

                                          Description  
0                              xDr session identifier  
1                   Total Duration of the xDR (in ms)  
2       Start time of the xDR (first frame timestamp)  
3   Milliseconds offset of start time for the xDR ...  
4          End time of the xDR (last frame timestamp)  
5   Milliseconds offset of end time of the xDR (la...  
6                    Total Duration of the xDR (in s)  
7            International Mobile Subscriber Identity  
8   MS International PSTN/ISDN Number of mobile - ...  
9             International Mobile Equipment Identity  
10  User location call name (2G/3G/4G) at the end ...  
11  Average Round Trip Time measurement Downlink d...  
12  Average Round Trip Time measurement Uplink dir...  
13  Average Bearer Throughput for Downlink (kbps) ...  
14  Average Bearer Throughput for uplink (kbps) - ...  
15  TCP volume of Downlink packets detected as ret...  
16  TCP volume of Uplink packets detected as retra...  
17  Duration ratio when Bearer Downlink Throughput...  
18  Duration ratio when Bearer Downlink Throughput...  
19  Duration ratio when Bearer Downlink Throughput...  
20  Duration ratio when Bearer Downlink Throughput...  
21  Duration ratio when Bearer Uplink Throughput < ….  
22  Duration 

Statistics

In [8]:
# number of elements in the df
tellco_data.size

8250055

In [10]:
# rows and columns in the df
tellco_data.shape

(150001, 55)

In [11]:
tellco_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150001 entries, 0 to 150000
Data columns (total 55 columns):
 #   Column                                    Non-Null Count   Dtype  
---  ------                                    --------------   -----  
 0   Bearer Id                                 149010 non-null  float64
 1   Start                                     150000 non-null  object 
 2   Start ms                                  150000 non-null  float64
 3   End                                       150000 non-null  object 
 4   End ms                                    150000 non-null  float64
 5   Dur. (ms)                                 150000 non-null  float64
 6   IMSI                                      149431 non-null  float64
 7   MSISDN/Number                             148935 non-null  float64
 8   IMEI                                      149429 non-null  float64
 9   Last Location Name                        148848 non-null  object 
 10  Avg RTT DL (ms)     

In [13]:
# descriptive statistics that summarize the central tendency, dispersion of the TellCodata's numerical columns, excluding NaN values
tellco_data.describe()

Bearer Id       Start ms         End ms     Dur. (ms)          IMSI  \
count  1.490100e+05  150000.000000  150000.000000  1.500000e+05  1.494310e+05   
mean   1.013887e+19     499.188200     498.800880  1.046086e+05  2.082016e+14   
std    2.893173e+18     288.611834     288.097653  8.103762e+04  2.148809e+10   
min    6.917538e+18       0.000000       0.000000  7.142000e+03  2.040471e+14   
25%    7.349883e+18     250.000000     251.000000  5.744050e+04  2.082014e+14   
50%    7.349883e+18     499.000000     500.000000  8.639900e+04  2.082015e+14   
75%    1.304243e+19     749.000000     750.000000  1.324302e+05  2.082018e+14   
max    1.318654e+19     999.000000     999.000000  1.859336e+06  2.140743e+14   

       MSISDN/Number          IMEI  Avg RTT DL (ms)  Avg RTT UL (ms)  \
count   1.489350e+05  1.494290e+05    122172.000000    122189.000000   
mean    4.188282e+10  4.847455e+13       109.795706        17.662883   
std     2.447443e+12  2.241637e+13       619.782739        84.793524   
min     3.360100e+10  4.400152e+11         0.000000         0.000000   
25%     3.365130e+10  3.546071e+13        32.000000         2.000000   
50%     3.366371e+10  3.572201e+13        45.000000         5.000000   
75%     3.368349e+10  8.611970e+13        70.000000        15.000000   
max     8.823971e+14  9.900120e+13     96923.000000      7120.000000   

       Avg Bearer TP DL (kbps)  ...  Youtube DL (Bytes)  Youtube UL (Bytes)  \
count            150000.000000  ...        1.500010e+05        1.500010e+05   
mean              13300.045927  ...        1.163407e+07        1.100941e+07   
std               23971.878541  ...        6.710569e+06        6.345423e+06   
min                   0.000000  ...        5.300000e+01        1.050000e+02   
25%                  43.000000  ...        5.833501e+06        5.517965e+06   
50%                  63.000000  ...        1.161602e+07        1.101345e+07   
75%               19710.750000  ...        1.744852e+07        1.651556e+07   
max              378160.000000  ...        2.325910e+07        2.201196e+07   

       Netflix DL (Bytes)  Netflix UL (Bytes)  Gaming DL (Bytes)  \
count        1.500010e+05        1.500010e+05       1.500010e+05   
mean         1.162685e+07        1.100175e+07       4.220447e+08   
std          6.725218e+06        6.359490e+06       2.439675e+08   
min          4.200000e+01        3.500000e+01       2.516000e+03   
25%          5.777156e+06        5.475981e+06       2.104733e+08   
50%          1.164222e+07        1.099638e+07       4.234081e+08   
75%          1.747048e+07        1.650727e+07       6.331742e+08   
max          2.325919e+07        2.201196e+07       8.434419e+08   

       Gaming UL (Bytes)  Other DL (Bytes)  Other UL (Bytes)  \
count       1.500010e+05      1.500010e+05      1.500010e+05   
mean        8.288398e+06      4.211005e+08      8.264799e+06   
std         4.782700e+06      2.432050e+08      4.769004e+06   
min         5.900000e+01      3.290000e+03      1.480000e+02   
25%         4.128476e+06      2.101869e+08      4.145943e+06   
50%         8.291208e+06      4.218030e+08      8.267071e+06   
75%         1.243162e+07      6.316918e+08      1.238415e+07   
max         1.655879e+07      8.434425e+08      1.655882e+07   

       Total UL (Bytes)  Total DL (Bytes)  
count      1.500000e+05      1.500000e+05  
mean       4.112121e+07      4.546434e+08  
std        1.127639e+07      2.441429e+08  
min        2.866892e+06      7.114041e+06  
25%        3.322201e+07      2.431068e+08  
50%        4.114331e+07      4.558411e+08  
75%        4.903424e+07      6.657055e+08  
max        7.833131e+07      9.029696e+08  

[8 rows x 50 columns]

It doesn't make sense to compute descriptive statistics for Beared Id, IMSI, MSISDN/Number and IMEI. But the describe method computes the statistics of all numeric columns. Another thing is these statisics are computed before the data is cleaned. So there might be changes after missing values and outliers are handled.

Cleaning Columns

In [14]:
tellco_data.shape

(150001, 55)

In [15]:
# number of rows in the description
len(tellco_field_desc_data)

56

There are 150001 rows and 55 columns in the dataframe. However, we have 56 columns with their names and descriptions in tellco_field_desc_df. This means there is one column that is described but not included in the dataframe.

In [16]:
comparison_df = pd.concat([ pd.Series(tellco_data.columns.tolist()), tellco_field_desc_data['Fields']], axis=1)
comparison_df.rename(columns={0: 'Column names in dataframe', 'Fields': 'Column names in description file'}, inplace=True)
comparison_df

Column names in dataframe  \
0                                  Bearer Id   
1                                      Start   
2                                   Start ms   
3                                        End   
4                                     End ms   
5                                  Dur. (ms)   
6                                       IMSI   
7                              MSISDN/Number   
8                                       IMEI   
9                         Last Location Name   
10                           Avg RTT DL (ms)   
11                           Avg RTT UL (ms)   
12                   Avg Bearer TP DL (kbps)   
13                   Avg Bearer TP UL (kbps)   
14               TCP DL Retrans. Vol (Bytes)   
15               TCP UL Retrans. Vol (Bytes)   
16                       DL TP < 50 Kbps (%)   
17            50 Kbps < DL TP < 250 Kbps (%)   
18             250 Kbps < DL TP < 1 Mbps (%)   
19                        DL TP > 1 Mbps (%)   
20                       UL TP < 10 Kbps (%)   
21             10 Kbps < UL TP < 50 Kbps (%)   
22            50 Kbps < UL TP < 300 Kbps (%)   
23                      UL TP > 300 Kbps (%)   
24                           HTTP DL (Bytes)   
25                           HTTP UL (Bytes)   
26                 Activity Duration DL (ms)   
27                 Activity Duration UL (ms)   
28                               Dur. (ms).1   
29                      Handset Manufacturer   
30                              Handset Type   
31           Nb of sec with 125000B < Vol DL   
32     Nb of sec with 1250B < Vol UL < 6250B   
33  Nb of sec with 31250B < Vol DL < 125000B   
34            Nb of sec with 37500B < Vol UL   
35    Nb of sec with 6250B < Vol DL < 31250B   
36    Nb of sec with 6250B < Vol UL < 37500B   
37             Nb of sec with Vol DL < 6250B   
38             Nb of sec with Vol UL < 1250B   
39                   Social Media DL (Bytes)   
40                   Social Media UL (Bytes)   
41                         Google DL (Bytes)   
42                         Google UL (Bytes)   
43                          Email DL (Bytes)   
44                          Email UL (Bytes)   
45                        Youtube DL (Bytes)   
46                        Youtube UL (Bytes)   
47                        Netflix DL (Bytes)   
48                        Netflix UL (Bytes)   
49                         Gaming DL (Bytes)   
50                         Gaming UL (Bytes)   
51                          Other DL (Bytes)   
52                          Other UL (Bytes)   
53                          Total UL (Bytes)   
54                          Total DL (Bytes)   
55                                       NaN   

            Column names in description file  
0                                  bearer id  
1                                  Dur. (ms)  
2                                      Start  
3                                   Start ms  
4                                        End  
5                                     End ms  
6                                   Dur. (s)  
7                                       IMSI  
8                              MSISDN/Number  
9                                       IMEI  
10                        Last Location Name  
11                           Avg RTT DL (ms)  
12                           Avg RTT UL (ms)  
13                   Avg Bearer TP DL (kbps)  
14                   Avg Bearer TP UL (kbps)  
15               TCP DL Retrans. Vol (Bytes)  
16               TCP UL Retrans. Vol (Bytes)  
17                       DL TP < 50 Kbps (%)  
18            50 Kbps < DL TP < 250 Kbps (%)  
19             250 Kbps < DL TP < 1 Mbps (%)  
20                        DL TP > 1 Mbps (%)  
21                       UL TP < 10 Kbps (%)  
22             10 Kbps < UL TP < 50 Kbps (%)  
23            50 Kbps < UL TP < 300 Kbps (%)  
24                      UL TP > 300 Kbps (%)  
25                           HTTP DL (Bytes)  
26           

In [17]:
tellco_data[['Dur. (ms)', 'Dur. (ms).1']]

Dur. (ms)   Dur. (ms).1
0       1823652.0  1.823653e+09
1       1365104.0  1.365104e+09
2       1361762.0  1.361763e+09
3       1321509.0  1.321510e+09
4       1089009.0  1.089009e+09
...           ...           ...
149996    81230.0  8.123076e+07
149997    97970.0  9.797070e+07
149998    98249.0  9.824953e+07
149999    97910.0  9.791063e+07
150000        NaN           NaN

[150001 rows x 2 columns]

Missing Values

In [57]:
percent_missing_values(tellco_data)

The dataset contains 12.72 % missing values.


In [58]:
missing_df = missing_values_table(tellco_data)

Your selected dataframe has 55 columns.
There are 41 columns that have missing values.


In [59]:
missing_df

Missing Values  % of Total Values  \
Nb of sec with 37500B < Vol UL                    130254              86.84   
Nb of sec with 6250B < Vol UL < 37500B            111843              74.56   
Nb of sec with 125000B < Vol DL                    97538              65.02   
TCP UL Retrans. Vol (Bytes)                        96649              64.43   
Nb of sec with 31250B < Vol DL < 125000B           93586              62.39   
Nb of sec with 1250B < Vol UL < 6250B              92894              61.93   
Nb of sec with 6250B < Vol DL < 31250B             88317              58.88   
TCP DL Retrans. Vol (Bytes)                        88146              58.76   
HTTP UL (Bytes)                                    81810              54.54   
HTTP DL (Bytes)                                    81474              54.32   
Avg RTT DL (ms)                                    27829              18.55   
Avg RTT UL (ms)                                    27812              18.54   
Handset Type                                        9559               6.37   
Handset Manufacturer                                9559               6.37   
Last Location Name                                  1153               0.77   
MSISDN/Number                                       1066               0.71   
Bearer Id                                            991               0.66   
Nb of sec with Vol UL < 1250B                        793               0.53   
10 Kbps < UL TP < 50 Kbps (%)                        792               0.53   
50 Kbps < UL TP < 300 Kbps (%)                       792               0.53   
UL TP < 10 Kbps (%)                                  792               0.53   
UL TP > 300 Kbps (%)                                 792               0.53   
Nb of sec with Vol DL < 6250B                        755               0.50   
DL TP > 1 Mbps (%)                                   754               0.50   
DL TP < 50 Kbps (%)                                  754               0.50   
250 Kbps < DL TP < 1 Mbps (%)                        754               0.50   
50 Kbps < DL TP < 250 Kbps (%)                       754               0.50   
IMEI                                                 572               0.38   
IMSI                                                 570               0.38   
Start                                                  1               0.00   
Start Time (ms)                                        1               0.00   
End                                                    1               0.00   
End Time (ms)                                          1               0.00   
Dur (s)                                                1               0.00   
Avg Bearer TP UL (kbps)                                1               0.00   
Avg Bearer TP DL (kbps)                                1               0.00   
Dur (ms)                                               1               0.00   
Activity Duration UL (ms)                              1               0.00   
Activity Duration DL (ms)                              1               0.00   
Total UL (Bytes)                                       1               0.00   
Total DL (Bytes)                                       1               0.00   

                                            Dtype  
Nb of sec with 37500B < Vol UL            float64  
Nb of sec with 6250B < Vol UL < 37500B    float64  
Nb of sec with 125000B < Vol DL           float64  
TCP UL Retrans. Vol (Bytes)               float64  
Nb of sec with 31250B < Vol DL < 125000B  float64  
Nb of sec with 1250B < Vol UL < 6250B     float64  
Nb of sec with 6250B < Vol DL < 31250B    float64  
TCP DL Retrans. Vol (Bytes)               float64  
HTTP UL (Bytes)                           float64  
HTTP DL (Bytes)                           float64  
Avg RTT DL (ms)                           float64  
Avg RTT UL (ms)                           float64  
Handset Type                               obje

It is good to remove columns with more than 30% of missing values.

In [60]:
columns_to_be_removed = missing_df[missing_df['% of Total Values'] >= 30.00].index.tolist()
columns_to_be_removed

['Nb of sec with 37500B < Vol UL',
 'Nb of sec with 6250B < Vol UL < 37500B',
 'Nb of sec with 125000B < Vol DL',
 'TCP UL Retrans. Vol (Bytes)',
 'Nb of sec with 31250B < Vol DL < 125000B',
 'Nb of sec with 1250B < Vol UL < 6250B',
 'Nb of sec with 6250B < Vol DL < 31250B',
 'TCP DL Retrans. Vol (Bytes)',
 'HTTP UL (Bytes)',
 'HTTP DL (Bytes)']

The TCP variables have too much missing values, instead of removing we will input since we need them for our analysis later.

In [50]:
columns_to_be_removed = [col for col in columns_to_be_removed if col not in ['TCP UL Retrans. Vol (Bytes)',
    'TCP DL Retrans. Vol (Bytes)']]
columns_to_be_removed

['Nb of sec with 37500B < Vol UL',
 'Nb of sec with 6250B < Vol UL < 37500B',
 'Nb of sec with 125000B < Vol DL',
 'Nb of sec with 31250B < Vol DL < 125000B',
 'Nb of sec with 1250B < Vol UL < 6250B',
 'Nb of sec with 6250B < Vol DL < 31250B',
 'HTTP UL (Bytes)',
 'HTTP DL (Bytes)']

In [17]:
tellco_clean_data = tellco_data.drop(columns_to_be_removed, axis=1)
tellco_clean_data.shape

(150001, 47)

checking the status of missing values in the modified dataframe.

In [62]:
percent_missing_values(tellco_clean_data)

The dataset contains 0.0 % missing values.


In [63]:
missing_values_table(tellco_clean_data)

Your selected dataframe has 47 columns.
There are 0 columns that have missing values.


Since the missing percentages of 'TCP UL Retrans. Vol (Bytes)' and 'TCP DL Retrans. Vol (Bytes)' are very high we will impute them with the bacward fill method. In this case, using a single value like mean or median is not advisable as it may heavily bais our data in an unwanted way by making the majority of values equal to a single value.

In [20]:
fix_missing_bfill(tellco_clean_data, 'TCP UL Retrans. Vol (Bytes)')

c:\Users\USER\Desktop\KAIM\TellCo-Data-Analysis\scripts\data_cleaner.py:93: FutureWarning: Series.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.
  df[col] = df[col].fillna(method='bfill')


96649 missing values in the column TCP UL Retrans. Vol (Bytes) have been replaced using the backward fill method.


0         7230.0
1         7230.0
2         7230.0
3         7230.0
4         7230.0
           ...  
149996       NaN
149997       NaN
149998       NaN
149999       NaN
150000       NaN
Name: TCP UL Retrans. Vol (Bytes), Length: 150001, dtype: float64

Avg RTT DL (ms) and Avg RTT UL (ms) have the next highest percentages of missing values with around 18.5% each.

In [22]:
tellco_clean_data['Avg RTT DL (ms)'].skew(skipna=True)

np.float64(62.90782807995961)

Since both the columns Avg RTT DL (ms) and Avg RTT UL (ms) are heavily positively skewed, it is advisable not to impute them with their mean. So we will use forward fill.

In [23]:
fix_missing_ffill(tellco_clean_data, 'Avg RTT DL (ms)')

27829 missing values in the column Avg RTT DL (ms) have been replaced using the forward fill method.


c:\Users\USER\Desktop\KAIM\TellCo-Data-Analysis\scripts\data_cleaner.py:86: FutureWarning: Series.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.
  df[col] = df[col].fillna(method='ffill')


0         42.0
1         65.0
2         65.0
3         65.0
4         65.0
          ... 
149996    32.0
149997    27.0
149998    43.0
149999    37.0
150000    37.0
Name: Avg RTT DL (ms), Length: 150001, dtype: float64

In [24]:
fix_missing_ffill(tellco_clean_data, 'Avg RTT UL (ms)')

27812 missing values in the column Avg RTT UL (ms) have been replaced using the forward fill method.


c:\Users\USER\Desktop\KAIM\TellCo-Data-Analysis\scripts\data_cleaner.py:86: FutureWarning: Series.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.
  df[col] = df[col].fillna(method='ffill')


0         5.0
1         5.0
2         5.0
3         5.0
4         5.0
         ... 
149996    0.0
149997    2.0
149998    6.0
149999    5.0
150000    5.0
Name: Avg RTT UL (ms), Length: 150001, dtype: float64

In [64]:
percent_missing_values(tellco_clean_data)

The dataset contains 0.0 % missing values.


In [26]:
percent_missing_rows(tellco_clean_data)

7.96 % of the rows in the dataset contain atleast one missing value.


In [51]:
missing_values_table(tellco_clean_data)

Your selected dataframe has 47 columns.
There are 0 columns that have missing values.


Since "Handset Type" and "Handset Manufacturer" are categorical columns, it is better to inpute them with the value "unknown" so that we don't bias the data.

In [28]:
fix_missing_value(tellco_clean_data, 'Handset Type', 'unknown')
fix_missing_value(tellco_clean_data, 'Handset Manufacturer', 'unknown')

9559 missing values in the column Handset Type have been replaced by unknown.
9559 missing values in the column Handset Manufacturer have been replaced by unknown.


0         Samsung
1         Samsung
2         Samsung
3         unknown
4         Samsung
           ...   
149996      Apple
149997      Apple
149998      Apple
149999     Huawei
150000    unknown
Name: Handset Manufacturer, Length: 150001, dtype: object

In [52]:
percent_missing_values(tellco_clean_data)

The dataset contains 0.0 % missing values.


In [30]:
percent_missing_rows(tellco_clean_data)

2.08 % of the rows in the dataset contain atleast one missing value.


Since only 2.08% of the rows contain atleast one missing value and the total number of rows is around 150000, dropping these rows won't have a noticable negative impact.

In [53]:
drop_rows_with_missing_values(tellco_clean_data)

0 rows containg missing values were dropped.


In [32]:
percent_missing_values(tellco_clean_data)

The dataset contains 0.0 % missing values.


Data Type Conversion

In [66]:
tellco_clean_data.dtypes

Bearer Id                                  int64
Start                             datetime64[ns]
Start ms                                 float64
End                               datetime64[ns]
End ms                                   float64
Dur. (ms)                                float64
IMSI                                       int64
MSISDN/Number                              int64
IMEI                                       int64
Last Location Name                string[python]
Avg RTT DL (ms)                          float64
Avg RTT UL (ms)                          float64
Avg Bearer TP DL (kbps)                  float64
Avg Bearer TP UL (kbps)                  float64
TCP DL Retrans. Vol (Bytes)              float64
TCP UL Retrans. Vol (Bytes)              float64
DL TP < 50 Kbps (%)                      float64
50 Kbps < DL TP < 250 Kbps (%)           float64
250 Kbps < DL TP < 1 Mbps (%)            float64
DL TP > 1 Mbps (%)                       float64
UL TP < 10 Kbps (%) 

In [34]:
tellco_clean_data

Bearer Id            Start  Start ms              End  End ms  \
0       1.311448e+19   4/4/2019 12:01     770.0  4/25/2019 14:35   662.0   
1       1.311448e+19   4/9/2019 13:04     235.0   4/25/2019 8:15   606.0   
2       1.311448e+19   4/9/2019 17:42       1.0  4/25/2019 11:58   652.0   
3       1.311448e+19   4/10/2019 0:31     486.0   4/25/2019 7:36   171.0   
4       1.311448e+19  4/12/2019 20:10     565.0  4/25/2019 10:40   954.0   
...              ...              ...       ...              ...     ...   
149991  7.349883e+18   4/29/2019 7:28     794.0   4/30/2019 0:36   523.0   
149992  1.311448e+19   4/29/2019 7:28     114.0   4/30/2019 7:04   724.0   
149993  1.311448e+19   4/29/2019 7:28      79.0  4/30/2019 18:22   512.0   
149994  1.311448e+19   4/29/2019 7:28      83.0  4/30/2019 15:01   268.0   
149995  1.304243e+19   4/29/2019 7:28     615.0   4/30/2019 0:01   407.0   

        Dur. (ms)          IMSI  MSISDN/Number          IMEI  \
0       1823652.0  2.082014e+14   3.366496e+10  3.552121e+13   
1       1365104.0  2.082019e+14   3.368185e+10  3.579401e+13   
2       1361762.0  2.082003e+14   3.376063e+10  3.528151e+13   
3       1321509.0  2.082014e+14   3.375034e+10  3.535661e+13   
4       1089009.0  2.082014e+14   3.369980e+10  3.540701e+13   
...           ...           ...            ...           ...   
149991    61661.0  2.082017e+14   3.376215e+10  3.586061e+13   
149992    84940.0  2.082014e+14   3.376127e+10  3.532701e+13   
149993   125622.0  2.082014e+14   3.362611e+10  3.573531e+13   
149994   113545.0  2.082003e+14   3.364566e+10  3.515541e+13   
149995    59587.0  2.082014e+14   3.366865e+10  3.533251e+13   

           Last Location Name  ...  Youtube DL (Bytes)  Youtube UL (Bytes)  \
0       9.16456699548519E+015  ...          15854611.0           2501332.0   
1                     L77566A  ...          20247395.0          19111729.0   
2                     D42335A  ...          19725661.0          14699576.0   
3                     T21824A  ...          21388122.0          15146643.0   
4                     D88865A  ...          15259380.0          18962873.0   
...                       ...  ...                 ...                 ...   
149991                D10033B  ...           7404819.0           9864285.0   
149992                D78058B  ...          22428728.0          14385815.0   
149993                D73542B  ...          18144703.0           8161828.0   
149994                T88383B  ...          23043782.0          19176074.0   
149995                T85721A  ...           6550499.0          18003146.0   

        Netflix DL (Bytes)  Netflix UL (Bytes)  Gaming DL (Bytes)  \
0                8198936.0           9656251.0        278082303.0   
1               18338413.0          17227132.0        608750074.0   
2               17587794.0           6163408.0        229584621.0   
3               13994646.0           1097942.0        799538153.0   
4               17124581.0            415218.0        527707248.0   
...                    ...                 ...                ...   
149991          18954727.0          19382787.0        258688968.0   
149992           6940672.0           6801943.0         35738570.0   
149993          20559422.0          20415237.0        569668054.0   
149994          18762809.0           9591310.0        810666072.0   
149995          22468983.0           7149728.0        784435351.0   

        Gaming UL (Bytes)  Other DL (Bytes)  Other UL (Bytes)  \
0              14344150.0       171744450.0         8814393.0   
1               1170709.0       526904238.0        15055145.0   
2                395630.0       410692588.0         4215763.0   
3              10849722.0       749039933.0        12797283.0   
4               3529801.0       550709500.0        13910322.0   
...                   ...               ...               ...   
149991          9542486.0       249358264.0        14134902.0   
149992          7842728.0       808482329.0

By observing the columns we can realize that "Start" and "End" columns are actually datetime values even though they are labelled  as objects by pandas. Besides these two cloumns all the other columns with object data types are actually string values. So we will convert these columns to their appropriate data types.

In [35]:
convert_to_datetime(tellco_clean_data, ['Start', 'End'])

In [36]:
# get the columns with object data type
string_columns = tellco_clean_data.select_dtypes(include='object').columns.tolist()
string_columns

['Last Location Name', 'Handset Manufacturer', 'Handset Type']

In [37]:
convert_to_string(tellco_clean_data, string_columns)

We also know that Bearer Id, IMSI, MSISDN/Number, IMEI are unique numbers used for identification. So for better readability we will change them from float64 to int64.

In [38]:
int_cols = ['Bearer Id', 'IMSI', 'MSISDN/Number', 'IMEI',]
convert_to_int(tellco_clean_data, int_cols)

In [67]:
tellco_clean_data.dtypes

Bearer Id                                  int64
Start                             datetime64[ns]
Start ms                                 float64
End                               datetime64[ns]
End ms                                   float64
Dur. (ms)                                float64
IMSI                                       int64
MSISDN/Number                              int64
IMEI                                       int64
Last Location Name                string[python]
Avg RTT DL (ms)                          float64
Avg RTT UL (ms)                          float64
Avg Bearer TP DL (kbps)                  float64
Avg Bearer TP UL (kbps)                  float64
TCP DL Retrans. Vol (Bytes)              float64
TCP UL Retrans. Vol (Bytes)              float64
DL TP < 50 Kbps (%)                      float64
50 Kbps < DL TP < 250 Kbps (%)           float64
250 Kbps < DL TP < 1 Mbps (%)            float64
DL TP > 1 Mbps (%)                       float64
UL TP < 10 Kbps (%) 

Dropping Duplicates

In [41]:
# search for duplicate rows and drop them
drop_duplicates(tellco_clean_data)

No duplicate rows were found.


As we have seen in the column cleaning section we have two duration columns one in seconds and the other in micro seconds. Let us check if the values are equal by converting the micro seconds to seconds.

In [69]:
print(tellco_clean_data.columns)

Index(['Bearer Id', 'Start', 'Start ms', 'End', 'End ms', 'Dur. (ms)', 'IMSI',
       'MSISDN/Number', 'IMEI', 'Last Location Name', 'Avg RTT DL (ms)',
       'Avg RTT UL (ms)', 'Avg Bearer TP DL (kbps)', 'Avg Bearer TP UL (kbps)',
       'TCP DL Retrans. Vol (Bytes)', 'TCP UL Retrans. Vol (Bytes)',
       'DL TP < 50 Kbps (%)', '50 Kbps < DL TP < 250 Kbps (%)',
       '250 Kbps < DL TP < 1 Mbps (%)', 'DL TP > 1 Mbps (%)',
       'UL TP < 10 Kbps (%)', '10 Kbps < UL TP < 50 Kbps (%)',
       '50 Kbps < UL TP < 300 Kbps (%)', 'UL TP > 300 Kbps (%)',
       'Activity Duration DL (ms)', 'Activity Duration UL (ms)', 'Dur. (ms).1',
       'Handset Manufacturer', 'Handset Type', 'Nb of sec with Vol DL < 6250B',
       'Nb of sec with Vol UL < 1250B', 'Social Media DL (Bytes)',
       'Social Media UL (Bytes)', 'Google DL (Bytes)', 'Google UL (Bytes)',
       'Email DL (Bytes)', 'Email UL (Bytes)', 'Youtube DL (Bytes)',
       'Youtube UL (Bytes)', 'Netflix DL (Bytes)', 'Netflix UL (Bytes)',


In [77]:
drop_columns(tellco_clean_data, ['Dur. (ms)'])

1 column was dropped.


Outlier Detection

In [78]:
columns = tellco_clean_data.select_dtypes('float64').columns.tolist()
columns

['Start ms',
 'End ms',
 'Avg RTT DL (ms)',
 'Avg RTT UL (ms)',
 'Avg Bearer TP DL (kbps)',
 'Avg Bearer TP UL (kbps)',
 'TCP DL Retrans. Vol (Bytes)',
 'TCP UL Retrans. Vol (Bytes)',
 'DL TP < 50 Kbps (%)',
 '50 Kbps < DL TP < 250 Kbps (%)',
 '250 Kbps < DL TP < 1 Mbps (%)',
 'DL TP > 1 Mbps (%)',
 'UL TP < 10 Kbps (%)',
 '10 Kbps < UL TP < 50 Kbps (%)',
 '50 Kbps < UL TP < 300 Kbps (%)',
 'UL TP > 300 Kbps (%)',
 'Activity Duration DL (ms)',
 'Activity Duration UL (ms)',
 'Dur. (ms).1',
 'Nb of sec with Vol DL < 6250B',
 'Nb of sec with Vol UL < 1250B',
 'Social Media DL (Bytes)',
 'Social Media UL (Bytes)',
 'Google DL (Bytes)',
 'Google UL (Bytes)',
 'Email DL (Bytes)',
 'Email UL (Bytes)',
 'Youtube DL (Bytes)',
 'Youtube UL (Bytes)',
 'Netflix DL (Bytes)',
 'Netflix UL (Bytes)',
 'Gaming DL (Bytes)',
 'Gaming UL (Bytes)',
 'Other DL (Bytes)',
 'Other UL (Bytes)',
 'Total UL (Bytes)',
 'Total DL (Bytes)']

In [85]:
tellco_clean_data['Social Media Data Volume (Bytes)'] = tellco_clean_data['Social Media UL (Bytes)'] + tellco_clean_data['Social Media DL (Bytes)']
tellco_clean_data['Google Data Volume (Bytes)'] = tellco_clean_data['Google UL (Bytes)'] + tellco_clean_data['Google DL (Bytes)']
tellco_clean_data['Email Data Volume (Bytes)'] = tellco_clean_data['Email UL (Bytes)'] + tellco_clean_data['Email DL (Bytes)']
tellco_clean_data['Youtube Data Volume (Bytes)'] = tellco_clean_data['Youtube UL (Bytes)'] + tellco_clean_data['Youtube DL (Bytes)']
tellco_clean_data['Netflix Data Volume (Bytes)'] = tellco_clean_data['Netflix UL (Bytes)'] + tellco_clean_data['Netflix DL (Bytes)']
tellco_clean_data['Gaming Data Volume (Bytes)'] = tellco_clean_data['Gaming UL (Bytes)'] + tellco_clean_data['Gaming DL (Bytes)']
tellco_clean_data['Other Data Volume (Bytes)'] = tellco_clean_data['Other UL (Bytes)'] + tellco_clean_data['Other DL (Bytes)']
tellco_clean_data['Total Data Volume (Bytes)'] = tellco_clean_data['Total UL (Bytes)'] + tellco_clean_data['Total DL (Bytes)']


In [86]:
tellco_clean_data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 146887 entries, 0 to 149995
Data columns (total 54 columns):
 #   Column                            Non-Null Count   Dtype         
---  ------                            --------------   -----         
 0   Bearer Id                         146887 non-null  int64         
 1   Start                             146887 non-null  datetime64[ns]
 2   Start ms                          146887 non-null  float64       
 3   End                               146887 non-null  datetime64[ns]
 4   End ms                            146887 non-null  float64       
 5   IMSI                              146887 non-null  int64         
 6   MSISDN/Number                     146887 non-null  int64         
 7   IMEI                              146887 non-null  int64         
 8   Last Location Name                146887 non-null  string        
 9   Avg RTT DL (ms)                   146887 non-null  float64       
 10  Avg RTT UL (ms)                   146

Saving data

In [87]:
# save the clean dataframe to a csv file
tellco_clean_data.to_csv('../data/cleaned_Tellco_data.csv')